### PR TITLE
Test Logger

### PR DESCRIPTION
**Pull request by bastbu**
# PR for issue #808

## What is being addressed

With this PR we introduce a logger which integrates with XUnit's `ITestOutputHelper` and use it for all integration tests. Using the new logger ensures that we get more log output when running tests from VS, and it can also be configured to print these logs to the console (should we want to use it in the CI).

image

### CODE REVIEW

File: Tests/Common/TestOutputLogger.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        public TestOutputLogger(ITestOutputHelper testOutputHelper) =>
+            this.testOutputHelper = testOutputHelper;
+
+        public IDisposable BeginScope<TState>(TState state) => NullDisposable.Instance;
+
+        public bool IsEnabled(LogLevel logLevel) => logLevel >= TestLogLevel;
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
+        {
+            if (!IsEnabled(logLevel)) return;
```
Comment (spygi) [inline Tests/Common/TestOutputLogger.cs:33]:
```suggestion
            _ = formatter ?? throw new ArgumentNullException(nameof(formatter));
            
            if (!IsEnabled(logLevel)) return;
```

**Review by spygi — APPROVED** [review]

File: Tests/Integration/ClassCIntegrationTests.cs [hunk 2]
```diff
@@ -15,14 +15,23 @@ namespace LoRaWan.Tests.Integration
...
     // End to end tests without external dependencies (IoT Hub, Service Facade Function)
     // Class C device tests
     public class ClassCIntegrationTests : MessageProcessorTestBase
     {
+        private readonly TestOutputLoggerFactory testOutputLoggerFactory;
+        private bool _disposedValue;
+
+        public ClassCIntegrationTests(ITestOutputHelper testOutputHelper) : base(testOutputHelper)
+        {
+            this.testOutputLoggerFactory = new TestOutputLoggerFactory(testOutputHelper);
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Tests/Integration/ClassCIntegrationTests.cs]: Done in bbcfdb0ff3d14c71f7a0a3de56deb4e3572d23b5

File: Tests/Common/TestOutputLogger.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+{
+    using System;
+    using System.Collections.Concurrent;
+    using LoRaWan.NetworkServer;
+    using Microsoft.Extensions.Logging;
+    using Xunit.Abstractions;
+
+    /// <summary>
+    /// Logger class that offers integration with XUnit's <see cref="ITestOutputHelper"/>.
+    /// It forwards log statements directly to <see cref="ITestOutputHelper"/> without taking into account scope information.
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Tests/Common/TestOutputLogger.cs:16]: Addressed in 03e0700d6b90a40590647e5ffeef38dc134ac389

File: Tests/Common/TestOutputLogger.cs (hunk 1 above)
Comment (bastbu) [reply Tests/Common/TestOutputLogger.cs:35]: Fixed the CA suggestion in de60a092a5a22f4aec3a8e689ae8cd1a55936d28

**Pull request merged**